File: 프로그래머스/dp/N으로표현.py
# number를 최소개의 N으로 표현.
# 최솟값이 8보다 크면 -1

# dp: N을 해당 index만큼 사용하여 만들 수 있는 모든 수  
# dp[1] 

def solution(N, number):
    dp = [0] + [set([int(str(N)*i)]) for i in range(1, 9)] # 1부터 8까지. 0번 index가 1, 7번 index가 8
    
    for i in range(1, 9):
      # a + b = i
      for j in range(1, i):
        k = i - j
        for a in dp[j]:
          for b in dp[k]:
            dp[i].add(a+b)
            dp[i].add(abs(a-b))
            dp[i].add(a*b)
            if b != 0:
              dp[i].add(int(a//b))
      
      if number in dp[i]:
        return i

        
solution(5, 12)
solution(2, 11)

File: 프로그래머스/dp/N으로표현.js
function solution(N, number) {
  if (N === number || N === 1) return (N % number) + 1;

  const visited = Array(320001).fill(0);
  const dp = Array.from({ length: 9 }, () => []);

  for (let i = 1; i < 6; i++) {
    const num = +("" + N).repeat(i);
    if (num <= 32000) {
      dp[i].push(num);
      visited[num] = i;
    }
  }

  const operators = ["+", "-", "/", "*"];
  const getNext = (a, b) => {
    const result = new Set();
    let res;
    for (let o of operators) {
      res1 = eval(a + o + b);
      res2 = eval(b + o + a);
      if (res1 >= 1 && res1 <= 32000) result.add(parseInt(res1));

      if (res2 >= 1 && res2 <= 32000) result.add(parseInt(res2));
    }
    return [...result];
  };

  for (let i = 2; i <= 8; i++) {
    for (let j = 1; j <= i / 2; j++) {
      for (let a of dp[j]) {
        for (let b of dp[i - j]) {
          for (let n of getNext(a, b)) {
            if (n === number) {
              if (i > 8) return -1;
              return i;
            }
            if (!visited[n]) {
              dp[i].push(n);
              visited[n] = i;
            }
          }
        }
      }
    }
  }

  return -1;
}
